heat_pump_adoption_modelling/pipeline/supervised_model/data_preprocessing.py

```diff
@@ -0,0 +1,584 @@
+from heat_pump_adoption_modelling.pipeline.encoding import feature_encoding
+from heat_pump_adoption_modelling.pipeline.preprocessing import feature_engineering
+
+import numpy as np
+import pandas as pd
+
+# ----------------------------------------------------------------------------------
+
+
+ordinal_features = [
+    "MAINHEAT_ENERGY_EFF",
+    "CURRENT_ENERGY_RATING",
+    "POTENTIAL_ENERGY_RATING",
+    "FLOOR_ENERGY_EFF",
+    "WINDOWS_ENERGY_EFF",
+    "HOT_WATER_ENERGY_EFF",
+    "LIGHTING_ENERGY_EFF",
+    "GLAZED_TYPE",
+    "MAINHEATC_ENERGY_EFF",
+    "WALLS_ENERGY_EFF",
+    "ROOF_ENERGY_EFF",
+    "MAINS_GAS_FLAG",
+    "CONSTRUCTION_AGE_BAND_ORIGINAL",
+    "CONSTRUCTION_AGE_BAND",
+    "N_ENTRIES",
+    "N_ENTRIES_BUILD_ID",
+    "ENERGY_RATING_CAT",
+]
+
+drop_features = [
+    "BUILDING_REFERENCE_NUMBER",
+    "ADDRESS1",
+    "ADDRESS2",
+    "POSTTOWN",
+    "LODGEMENT_DATE",
+    "CO2_EMISS_CURR_PER_FLOOR_AREA",
+    "MAINHEAT_DESCRIPTION",
+    "SHEATING_ENERGY_EFF",
+    "HEATING_COST_POTENTIAL",
+    "HOT_WATER_COST_POTENTIAL",
+    "LIGHTING_COST_POTENTIAL",
+    "CONSTRUCTION_AGE_BAND",
+    "NUMBER_HEATED_ROOMS",
+    "LOCAL_AUTHORITY_LABEL",
+    "ENTRY_YEAR",
+    "N_ENTRIES",
+    "CURR_ENERGY_RATING_NUM",
+    "ENERGY_RATING_CAT",
+    "UNIQUE_ADDRESS",
+    "INSPECTION_DATE",
+    "MAIN_FUEL",
+    "Country",
+    "original_address",
+    "FIRST_HP_MENTION",
+    "INSPECTION_YEAR",
+    "HP_INSTALL_YEAR",
+    "FIRST_HP_MENTION_YEAR",
+    "MCS_AVAILABLE",
+    "HAS_HP_AT_SOME_POINT",
+    "HP_TYPE",
+]
+
+# Load config file
+config = get_yaml_config(
+    Path(str(PROJECT_DIR) + "/heat_pump_adoption_modelling/config/base.yaml")
+)
+
+# Get paths
+MERGED_MCS_EPC = str(PROJECT_DIR) + config["MERGED_MCS_EPC"]
+SUPERVISED_MODEL_OUTPUT = str(PROJECT_DIR) + config["SUPERVISED_MODEL_OUTPUT"]
+SUPERVISED_MODEL_FIG_PATH = str(PROJECT_DIR) + config["SUPERVISED_MODEL_FIG_PATH"]
+dtypes = config["dtypes"]
+
+
+def select_samples_by_postcode_completeness(
+    df, min_samples=5000000, start=250000, interval=1000
+):
+    """Select a subset of samples with a minimal number of samples,
+    including all samples for each unique postcode.
+    This is done by increasing the number of postcodes and adding up the samples.
+
+    Parameters
+    ----------
+    df: pandas.Dataframe
+        Dataframe that includes the samples from which to select subet.
+
+    min_samples: int, default=5000000
+        Number of minimal samples, stop after reaching this number.
+
+    start: int, default=25000
+        Start with this number of postcodes.
+
+    interval: int, default=1000
+        The interval by which to increase number of postcodes every iteration.
+
+    Return
+    ---------
+    df_reduced : pandas.Dataframe
+        Subset of original dataframe
+        with minimal number of samples and complete postcodes."""
+
+    # Iterate over batches of postcodes and add up samples
+    for i in range(start, min_samples, interval):
```

Comment:
Thanks a lot! I knew this wasn't ideal but since I only ran it once, I didn't invest too much. 
But you're absolutely right, your suggestion is much faster and cleaner. Implemented now.